### PR TITLE
batches: issue executor warning when outdated

### LINKED ISSUE #27570: executors: Warn when outdated
We currently rely on the executors being in version sync with the backend, so we should either not allow starting up when the version doesn't match a specific version constraint or at least warn site-admins that the instances are outdated.

### PR DESCRIPTION
**Pull request by BolajiOlajide**
Closes #27570

## Test plan

* Confirm warning only shows up on active executors that have a version lower than the sourcegraph version (either using semver versioning or timestamps).

CleanShot 2022-08-26 at 16 02 13@2x

## App preview:

- [Web](https://sg-web-bo-outdated-executor-warning.onrender.com)
- [Storybook](https://5f0f381c0e50750022dc6bf7-fluycnvdxv.chromatic.com)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

**Comment by BolajiOlajide** [comment]
@danielmarquespt Is the wording for this alert okay?
CleanShot 2022-08-26 at 16 02 13@2x

**Comment by danielmarquespt** [comment]
> @danielmarquespt Is the wording for this alert okay?
> CleanShot 2022-08-26 at 16 02 13@2x

@BolajiOlajide Are able to tell the user what would be a compatible version?

**Comment by BolajiOlajide** [comment]
> > @danielmarquespt Is the wording for this alert okay?
> > CleanShot 2022-08-26 at 16 02 13@2x
> 
> @BolajiOlajide Are able to tell the user which version would be a compatible version?

Yes.

**Comment by danielmarquespt** [comment]
> > > @danielmarquespt Is the wording for this alert okay?
> > > CleanShot 2022-08-26 at 16 02 13@2x
> > 
> > 
> > @BolajiOlajide Are able to tell the user which version would be a compatible version?
> 
> Yes.

Then I would suggest something like "Please upgrade this executor to a version compatible with your Sourcegraph installation (X.X or higher)"

File: client/web/src/enterprise/batches/batch-spec/execute/TimelineModal.tsx [hunk 1]
```diff
@@ -21,26 +30,44 @@ export interface TimelineModalProps {
...
+        return <LoadingSpinner />
+    }
+
+    if (error || !data?.site) {
+        const title = error ? String(error) : 'Unable to fetch sourcegraph version.'
+        return <HeroPage icon={AlertCircleIcon} title={title} />
+    }
+
+    return (
+        <Modal className={styles.modalBody} position="center" onDismiss={onCancel} aria-label="Execution timeline">
```
Comment (eseliger) [inline client/web/src/enterprise/batches/batch-spec/execute/TimelineModal.tsx:51]: I think this causes the modal to not render until the request is done, right? That's kinda weird from a UX perspective as it means the button press on "Timeline" doesn't immediately do anything.

I think this wouldn't be an issue if the logic for this was on the backend.

File: client/web/src/enterprise/executors/utils.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+ * 169135_2022-08-25_a2b623dce148
+ * 169120_2022-08-25_a94c7eb7beca
+ *
+ * Valid build date example for executor (patch)
+ * executor-patch-notest-es-ignite-debug_168065_2022-08-18_e94e18c4ebcc_patch
+ */
+const buildDateRegex = /^[\w-]+_(\d{4}-\d{2}-\d{2})_\w+/
+const developmentVersion = '0.0.0+dev'
+
+export const isExecutorVersionOutdated = (
```
Comment (eseliger) [inline client/web/src/enterprise/executors/utils.ts:17]: Instead of encapsulating this knowledge in the frontend, I would personally opt for putting it in the backend and adding a resolver field `outdated`. 
Reasons:
- If someone outside of Sourcegraph wanted to use this information, they would have to re-build this logic
- If we wanted to add some tooling or check to validate there are never outdated executors connected, we would need to reimplement it as well
- There's a similar logic piece in src-cli for version matching in go, that might be possible to re-use. 

WDYT?

Comment (BolajiOlajide) [reply client/web/src/enterprise/executors/utils.ts:17]: Makes sense to me.

Comment (courier-new) [reply client/web/src/enterprise/executors/utils.ts:17]: Agreed, but I do love all the tests you wrote for this. 😄

File: client/web/src/enterprise/batches/batch-spec/execute/TimelineModal.tsx (hunk 1 above)
Comment (BolajiOlajide) [reply client/web/src/enterprise/batches/batch-spec/execute/TimelineModal.tsx]: Moved the logic to the backend.

File: client/web/src/enterprise/executors/ExecutorsListPage.tsx [hunk 3]
```diff
@@ -206,6 +206,15 @@ export const ExecutorNode: FunctionComponent<React.PropsWithChildren<ExecutorNod
                 </div>
             </dl>
         </Collapsible>
+
+        {node.isOutdated && (
+            <Alert variant="warning" className="mt-3">
+                <Text className="m-0">{node.hostname} is outdated.</Text>
+                <Text className="m-0">
+                    Please upgrade this executor to a version compatible with your Sourcegraph version.
```
[the remark this answers is not in the document]
Comment (BolajiOlajide) [reply client/web/src/enterprise/executors/ExecutorsListPage.tsx:214]: This I believe takes care of this condition. If the executor is too new, the executor version in the `executor_heartbeats` table should be an empty string.

File: internal/services/executors/transport/graphql/executor_resolver.go [hunk 4]
```diff
@@ -39,6 +42,45 @@ func (e *ExecutorResolver) SrcCliVersion() string   { return e.executor.SrcCliVe
...
+
+	if !active || isSgDev || isExecutorDev {
+		return false, nil
+	}
+
+	r := regexp.MustCompile(`^[\w-]+_(\d{4}-\d{2}-\d{2})_\w+`)
+	evm := r.FindStringSubmatch(ev)
+	svm := r.FindStringSubmatch(sv)
+	if len(evm) > 1 && len(svm) > 1 {
+		return svm[1] > evm[1], nil
```
[the remark this answers is not in the document]
Comment (BolajiOlajide) [reply internal/services/executors/transport/graphql/executor_resolver.go:64]: Lol. No worries.
Figured an approach that's better [here](https://go.dev/play/p/RQbjr0EGX2_9).

File: internal/services/executors/transport/graphql/executor_resolver.go [hunk 5]
```diff
@@ -39,6 +42,45 @@ func (e *ExecutorResolver) SrcCliVersion() string   { return e.executor.SrcCliVe
...
+	}
+
+	// if we get here then we assume the versions are in semver format
+	// we use the sourcegraph version as a constraint and expect the executor version to be
+	// greater than the sourcegraph version, if not we return true to indicate the executor is
+	// outdated.
+	c, err := semver.NewConstraint("< " + sv)
+	if err != nil {
+		return true, err
+	}
```
[the remark this answers is not in the document]
Comment (BolajiOlajide) [reply internal/services/executors/transport/graphql/executor_resolver.go:74]: I'm guessing you mean one minor version ahead.

[the remark this answers is not in the document]
Comment (eseliger) [reply internal/services/executors/transport/graphql/executor_resolver.go]: yes, sorry. Or technically speaking, 3.43 and 4.0 are also compatible, to make things harder :)

File: client/web/src/enterprise/executors/ExecutorsListPage.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (eseliger) [reply client/web/src/enterprise/executors/ExecutorsListPage.tsx:214]: how is it empty? Maybe it's super obvious but I'm not seeing it rn, sorry 😬

File: internal/services/executors/transport/graphql/executor_resolver.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (BolajiOlajide) [reply internal/services/executors/transport/graphql/executor_resolver.go]: Yeah, however, this (4.0 and 3.43 being compatible) will be tricky, because I'm not sure there's a clean way to know that `3.43` is the last minor version under the `3.x.x` release.

I can attempt this by saying that if the major version is greater and the minor is zero then assume the executor isn't outdated, then the problem now would be that `3.40.0` will be seen as compatible with `4.0`

File: client/web/src/enterprise/executors/ExecutorsListPage.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (BolajiOlajide) [reply client/web/src/enterprise/executors/ExecutorsListPage.tsx:214]: Oh no, it's fine, I noticed that when an executor is new, the version isn't stored in the database until after a couple of seconds.
I recorded[ a loom ](https://www.loom.com/share/4e1fbaf19053402bb1cb027ba6bfb016)to show this.

File: client/web/src/enterprise/executors/ExecutorsListPage.tsx [hunk 6]
```diff
@@ -50,7 +50,7 @@ export const ExecutorsListPage: FunctionComponent<React.PropsWithChildren<Execut
     queryExecutors = defaultQueryExecutors,
     ...props
 }) => {
-    useEffect(() => eventLogger.logViewEvent('ExecutorsList'))
+    useEffect(() => eventLogger.logPageView('ExecutorsList'))
```
Comment (courier-new) [inline client/web/src/enterprise/executors/ExecutorsListPage.tsx:53]: Is this like a deprecated API thing? I feel like I've seen a warning about this before. Should we file a ticket to update across the codebase on a rainy day? 🙂

File: client/web/src/enterprise/executors/ExecutorsListPage.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (courier-new) [reply client/web/src/enterprise/executors/ExecutorsListPage.tsx:214]: Interesting, that feels... unintentional? 😅 Anyway, does that mean `isExecutorOutdated` compares with `0.0.0`, or something different?

File: internal/services/executors/transport/graphql/executor_resolver.go [hunk 7]
```diff
@@ -39,6 +42,56 @@ func (e *ExecutorResolver) SrcCliVersion() string   { return e.executor.SrcCliVe
 func (e *ExecutorResolver) FirstSeenAt() DateTime   { return DateTime{e.executor.FirstSeenAt} }
 func (e *ExecutorResolver) LastSeenAt() DateTime    { return DateTime{e.executor.LastSeenAt} }
 
+func (e *ExecutorResolver) IsOutdated() (bool, error) {
+	ev := e.executor.ExecutorVersion
+	return isExecutorOutdated(ev, e.Active())
```
Comment (courier-new) [inline internal/services/executors/transport/graphql/executor_resolver.go:47]: Maybe should an inactive executor just early return here?

```suggestion
    if !e.Active() {
        return false, nil
    }
	return isExecutorOutdated(ev)
```

It feels unrelated to the version check and at first I wondered if there was a different relationship between the version and whether or not the executor is active. 😅

File: CHANGELOG.md [hunk 8]
```diff
@@ -18,6 +18,7 @@ All notable changes to Sourcegraph are documented in this file.
 ### Added
 
 - A new look for Sourcegraph, previously in beta as "Simple UI", is now permanently enabled. [#41021](https://github.com/sourcegraph/sourcegraph/pull/41021)
+- Display warning when executor is outdated. [#40916](https://github.com/sourcegraph/sourcegraph/pull/40916)
```
Comment (courier-new) [inline CHANGELOG.md:21]:
```suggestion
- Outdated executors now show a warning from the admin page. [#40916](https://github.com/sourcegraph/sourcegraph/pull/40916)
```

File: client/web/src/enterprise/executors/ExecutorsListPage.tsx (hunk 3 above)
Comment (courier-new) [inline client/web/src/enterprise/executors/ExecutorsListPage.tsx:214]: Could we put a link to the executor docs most relevant to upgrading here on the word "upgrade"?

File: internal/services/executors/transport/graphql/executor_resolver.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (courier-new) [reply internal/services/executors/transport/graphql/executor_resolver.go]: I think we just take a loss here and false-positive display the warning with 3.43 & 4.0. Not that we're following semver conventions particularly closely with our 4.0 release, but in most systems there's no guarantee of compatibility between major versions, so it wouldn't feel unexpected to a user if the banner appears when they upgrade SG to 4.0. Worst case, someone upgrades their executors sooner. 🙂

**Review by courier-new — APPROVED** [review]

File: client/web/src/enterprise/executors/ExecutorsListPage.tsx (hunk 3 above)
Comment (eseliger) [reply client/web/src/enterprise/executors/ExecutorsListPage.tsx:214]: cute, she thinks we have good upgrade docs :lolsob:

File: internal/services/executors/transport/graphql/executor_resolver.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (eseliger) [reply internal/services/executors/transport/graphql/executor_resolver.go]: yeah that's probably cool for now :)

File: client/web/src/enterprise/executors/ExecutorsListPage.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (eseliger) [reply client/web/src/enterprise/executors/ExecutorsListPage.tsx:214]: Ah yeah that's a side-effect of that being the`executor_heartbeats` table, not the executors and their stats. It will only be populated once we receive a heartbeat. I'll take a note of it but it's sort of "expected" for now.

[the remark this answers is not in the document]
Comment (eseliger) [reply client/web/src/enterprise/executors/ExecutorsListPage.tsx:214]: but then once it's saved, it can still be newer than Sourcegraph, no?

[the remark this answers is not in the document]
Comment (BolajiOlajide) [reply client/web/src/enterprise/executors/ExecutorsListPage.tsx:214]: When you say newer what do you mean?

[the remark this answers is not in the document]
Comment (eseliger) [reply client/web/src/enterprise/executors/ExecutorsListPage.tsx:214]: The executor is version 4.11 but sourcegraph is 3.39

[the remark this answers is not in the document]
Comment (BolajiOlajide) [reply client/web/src/enterprise/executors/ExecutorsListPage.tsx:214]: Oh, then I totally misunderstood your initial comment, my bad!
Ignore everything I've said so far lol.

If the executor is newer, then I agree the wording doesn't explain that properly. However, the current logic implemented in this PR won't flag that as an outdated executor since technically it's the Sourcegraph instance that's outdated. I don't think that should change imo.

File: client/web/src/enterprise/executors/ExecutorsListPage.tsx (hunk 6 above)
Comment (BolajiOlajide) [reply client/web/src/enterprise/executors/ExecutorsListPage.tsx:53]: Yes, it is a deprecated method. https://sourcegraph.sourcegraph.com/github.com/sourcegraph/sourcegraph@main/-/blob/client/web/src/tracking/eventLogger.ts?L120%3A1-132%3A6=&utm_source=VSCode-2.2.9

Created a [PR for the update](https://github.com/sourcegraph/sourcegraph/pull/41354).

File: cmd/frontend/graphqlbackend/schema.graphql [hunk 9]
```diff
@@ -1704,6 +1704,11 @@ type Executor implements Node {
     The last time the executor sent a heartbeat to the Sourcegraph instance.
     """
     lastSeenAt: DateTime!
+    """
+    True, if the executor does not match the version constraint of the Sourcegraph installation.
+    If outdated, please make sure that the executor and the Sourcegraph backend are of compatible versions. This means they should match in major and minor version, but they may be 1 minor version apart.
+    """
+    isOutdated: Boolean!
```
Comment (Piszmog) [inline cmd/frontend/graphqlbackend/schema.graphql:1711]: Based on the conversation about the wording in the alert, is the name of this field misleading? "Outdated" to me would suggest that the executor can only be behind the version of Sourcegraph. But the conversation about the alert would suggest that the executor can be ahead of Sourcegraph.

Would something like `isIncompatible` or `versionMismatch` be better?

**Review by Piszmog — APPROVED** [review]

File: client/web/src/enterprise/executors/ExecutorsListPage.tsx (hunk 6 above)
Comment (courier-new) [reply client/web/src/enterprise/executors/ExecutorsListPage.tsx:53]: WOW even better, thanks Bolaji!

File: client/web/src/enterprise/executors/ExecutorsListPage.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (BolajiOlajide) [reply client/web/src/enterprise/executors/ExecutorsListPage.tsx:214]: > Should we cover both cases?

Yes, to cover both cases.

> We can do it in two separate PRs though if you want.

I'm thinking of what that'd look like before deciding if it should be a separate PR or not. 
Ideally, I'll prefer to split it up. Thinking if adding a new field to the schema is the best approach.

File: cmd/frontend/graphqlbackend/schema.graphql (hunk 9 above)
Comment (BolajiOlajide) [reply cmd/frontend/graphqlbackend/schema.graphql:1711]: Yes, I pondered a bit about this earlier today. I think having a separate field here should work but not sure if it's worth it. Since it's a boolean field, it's a bit tricky because there are more than two possible states.

* Executor is outdated
* Executor is not outdated
* Executor is too new

Comment (Piszmog) [reply cmd/frontend/graphqlbackend/schema.graphql:1711]: Or have an enum for the different states?

File: client/web/src/enterprise/executors/ExecutorsListPage.tsx (hunk 6 above)
Comment (BolajiOlajide) [reply client/web/src/enterprise/executors/ExecutorsListPage.tsx:53]: I'm going to revert this change temporarily. There needs to be some form of data migration before doing this and the PR I shared above already takes care of it.

File: cmd/frontend/graphqlbackend/schema.graphql [hunk 10]
```diff
@@ -1918,6 +1924,26 @@ enum SearchPatternType {
...
+
+    """
+    Executor is up-to-date with the sourcegraph instance.
+    """
+    UPTODATE
+
+    """
+    Executor version is more than one version ahead of the sourcegraph instance.
+    """
+    TOONEW
```
Comment (BolajiOlajide) [inline cmd/frontend/graphqlbackend/schema.graphql:1944]: Does anyone have ideas for what this should be? I've run out of ideas for what to name this enum field.

cc @courier-new @danielmarquespt @eseliger @Piszmog

File: internal/services/executors/transport/graphql/executor_resolver.go [hunk 11]
```diff
@@ -39,6 +42,75 @@ func (e *ExecutorResolver) SrcCliVersion() string   { return e.executor.SrcCliVe
 func (e *ExecutorResolver) FirstSeenAt() DateTime   { return DateTime{e.executor.FirstSeenAt} }
 func (e *ExecutorResolver) LastSeenAt() DateTime    { return DateTime{e.executor.LastSeenAt} }
 
+func (e *ExecutorResolver) Compatibility() (string, error) {
```
Comment (BolajiOlajide) [inline internal/services/executors/transport/graphql/executor_resolver.go:45]: Renamed this to `Compatibility`, I think that covers version mismatch (backward or forward).

File: cmd/frontend/graphqlbackend/schema.graphql (hunk 10 above)
Comment (eseliger) [reply cmd/frontend/graphqlbackend/schema.graphql:1944]: Hmm. I don't _hate_ `TOO_NEW` fwiw, but otherwise `VERSION_AHEAD`?

Comment (Piszmog) [reply cmd/frontend/graphqlbackend/schema.graphql:1944]: I like `VERSION_AHEAD`

File: internal/services/executors/transport/graphql/executor_resolver.go [hunk 12]
```diff
@@ -39,6 +42,75 @@ func (e *ExecutorResolver) SrcCliVersion() string   { return e.executor.SrcCliVe
...
+	e, err := semver.NewVersion(ev)
+	if err != nil {
+		return compatibility.ToGraphQL(), err
+	}
+
+	// it's okay for an executor to be one version behind or ahead of the sourcegraph version.
+	iev := e.IncMinor()
+
+	isv := s.IncMinor()
+	// dev, err := semver.NewVersion(fmt.Sprintf("%d.%d.%d", e.Major(), e.Minor() - 1, e.Patch()))
```
[the remark this answers is not in the document]
Comment (BolajiOlajide) [reply internal/services/executors/transport/graphql/executor_resolver.go:103]: Dang. My bad. 
Thanks for catching that.

File: client/web/src/enterprise/executors/ExecutorsListPage.tsx [hunk 13]
```diff
@@ -120,94 +121,131 @@ export interface ExecutorNodeProps {
...
-                            {node.active ? (
-                                <Icon
-                                    aria-hidden={true}
-                                    className="text-success mr-2"
-                                    svgPath={mdiCheckboxBlankCircle}
-                                />
-                            ) : (
-                                <Tooltip content="This executor missed at least three heartbeats.">
+export const ExecutorNode: FunctionComponent<React.PropsWithChildren<ExecutorNodeProps>> = ({ node }) => {
+    const renderAlert = (): JSX.Element | null => {
```
Comment (eseliger) [inline client/web/src/enterprise/executors/ExecutorsListPage.tsx:125]: This is a de facto react function component, can we make it a proper one by pulling it out of here and declaring it as a React.FunctionComponent?

File: client/web/src/enterprise/executors/ExecutorsListPage.tsx [hunk 14]
```diff
@@ -120,94 +121,131 @@ export interface ExecutorNodeProps {
...
+                                upgrade this executor
+                            </Link>{' '}
+                            to a version compatible with your Sourcegraph version.
+                        </Text>
+                    </Alert>
+                )
+            case ExecutorCompatibility.VERSION_AHEAD:
+                return (
+                    <Alert variant="warning" className="mt-3">
+                        <Text className="m-0">Your sourcegraph instance is out of date.</Text>
```
Comment (eseliger) [inline client/web/src/enterprise/executors/ExecutorsListPage.tsx:143]:
```suggestion
                        <Text className="m-0">Your Sourcegraph instance is out of date.</Text>
```

File: client/web/src/enterprise/executors/ExecutorsListPage.tsx [hunk 15]
```diff
@@ -120,94 +121,131 @@ export interface ExecutorNodeProps {
...
+                        <Text className="m-0">
+                            Please{' '}
+                            <Link to="/help/admin/updates" target="_blank" rel="noopener">
+                                upgrade your sourcegraph instance
```
Comment (eseliger) [inline client/web/src/enterprise/executors/ExecutorsListPage.tsx:147]:
```suggestion
                                upgrade your Sourcegraph instance
```

Comment (eseliger) [reply client/web/src/enterprise/executors/ExecutorsListPage.tsx:147]: should we only recommend this or also that downgrading executors is an option?

File: internal/services/executors/transport/graphql/executor_resolver_test.go [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+				isActive:           false,
+				expected:           UpToDateCompatibility.ToGraphQL(),
+			},
+			// The executor is not outdated if it's one minor version ahead of the sourcegraph version (SEMVER)
+			{
+				executorVersion:    "3.42.0",
+				sourcegraphVersion: "3.43.0",
+				isActive:           false,
+				expected:           UpToDateCompatibility.ToGraphQL(),
+			},
```
Comment (eseliger) [inline internal/services/executors/transport/graphql/executor_resolver_test.go:47]: would you mind adding a 4.0.0 -> 3.43.0, 3.43.0 -> 4.0.0 case here, just so that we know what it returns?

File: internal/services/executors/transport/graphql/executor_resolver_test.go [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+			{
+				executorVersion:    "3.43.0",
+				sourcegraphVersion: "3.40.0",
+				isActive:           true,
+				expected:           VersionAheadCompatibility.ToGraphQL(),
+			},
+			// The executor is outdated if the sourcegraph version is greater than the executor version (BuildDate).
+			{
+				executorVersion:    "executor-patch-notest-es-ignite-debug_168065_2022-08-20_e94e18c4ebcc_patch",
+				sourcegraphVersion: "169135_2022-08-25_a2b623dce148",
```
Comment (eseliger) [inline internal/services/executors/transport/graphql/executor_resolver_test.go:80]: what does a mix of 
(dev, tagged version)
(insiders tag, tagged version)
(insiders tag, dev)
... etc yield here?

**Review by eseliger — COMMENTED** [review]
sorry for the slow review! Added a few questions inline

File: internal/services/executors/transport/graphql/executor_resolver_test.go (hunk 16 above)
Comment (BolajiOlajide) [reply internal/services/executors/transport/graphql/executor_resolver_test.go]: Sure.

File: internal/services/executors/transport/graphql/executor_resolver_test.go (hunk 17 above)
Comment (BolajiOlajide) [reply internal/services/executors/transport/graphql/executor_resolver_test.go:92]: A mix of `dev` against any other kind of version will result in `UP_TO_DATE` compatibility. 
Having the [API return null](https://github.com/sourcegraph/sourcegraph/pull/40916#discussion_r976669327) could be a better approach for this mix of versioning.

Regarding the insiders version mix, I didn't handle that as I was unaware we had insiders versioning for executor instances. Having the API return null for dev / mismatch versions / insiders version should work fine, WDYT?

File: client/web/src/enterprise/executors/ExecutorsListPage.tsx (hunk 15 above)
Comment (BolajiOlajide) [reply client/web/src/enterprise/executors/ExecutorsListPage.tsx]: Okay, I can add that. I'll add a link to [this](https://docs.sourcegraph.com/admin/deploy_executors) for downgrading executors.

File: internal/services/executors/transport/graphql/executor_resolver_test.go (hunk 17 above)
Comment (eseliger) [reply internal/services/executors/transport/graphql/executor_resolver_test.go:92]: can we add that as a test case, just so we preserve this behavior?

yeah that's ok

File: internal/services/executors/transport/graphql/executor_resolver.go [hunk 18]
```diff
@@ -39,6 +42,88 @@ func (e *ExecutorResolver) SrcCliVersion() string   { return e.executor.SrcCliVe
...
+		return nil, nil
+	}
+
+	r := regexp.MustCompile(`^[\w-]+_(\d{4}-\d{2}-\d{2})_\w+`)
+	evm := r.FindStringSubmatch(ev)
+	svm := r.FindStringSubmatch(sv)
+
+	// check for version mismatch
+	if len(evm) > 1 && len(svm) <= 1 {
+		// this means that the executor is an insiert version while the Sourcegraph
```
Comment (LawnGnome) [inline internal/services/executors/transport/graphql/executor_resolver.go:70]:
```suggestion
		// this means that the executor is an insider version while the Sourcegraph
```

File: cmd/frontend/graphqlbackend/schema.graphql [hunk 19]
```diff
@@ -1730,6 +1729,12 @@ type Executor implements Node {
     The last time the executor sent a heartbeat to the Sourcegraph instance.
     """
     lastSeenAt: DateTime!
+    """
+    The compatibility of the executor with respect to the Sourcegraph instance.
+    If outdated, please make sure that the executor and the Sourcegraph backend are of compatible versions. This means they should match in major and minor version, but they may be 1 minor version apart.
+    If too new, please update the sourcegraph instance to match the version of the executor.
+    """
```
Comment (LawnGnome) [inline cmd/frontend/graphqlbackend/schema.graphql:1736]: Since this is nullable, we should also document what a `null` return value represents. (Which, reading the resolver code, I know is basically "we have no idea, good luck", but let's put it a little more nicely than that!)

File: internal/services/executors/transport/graphql/executor_resolver.go [hunk 20]
```diff
@@ -39,6 +42,88 @@ func (e *ExecutorResolver) SrcCliVersion() string   { return e.executor.SrcCliVe
...
+	sv := version.Version()
+
+	isExecutorDev := ev != "" && version.IsDev(ev)
+	isSgDev := sv != "" && version.IsDev(sv)
+
+	if isSgDev || isExecutorDev {
+		return nil, nil
+	}
+
+	r := regexp.MustCompile(`^[\w-]+_(\d{4}-\d{2}-\d{2})_\w+`)
```
Comment (LawnGnome) [inline internal/services/executors/transport/graphql/executor_resolver.go:64]: Let's move this out to a file-level variable so that, if it somehow panics, that happens at startup and not at a random later stage. (Also, saves us recompiling each time.)

**Review by LawnGnome — APPROVED** [review]
:+1:

File: client/shared/src/commandPalette/CommandList.tsx [hunk 21]
```diff
@@ -416,7 +416,7 @@ export const CommandListPopoverButton = forwardRef((props, ref) => {
                 ref={ref}
                 // Support legacy buttonElement prop since it's used in the different code hosts
                 // specifications
-                as={buttonElement as 'button' ?? Component}
+                as={(buttonElement as 'button') ?? Component}
```
[the remark this answers is not in the document]
Comment (BolajiOlajide) [reply client/shared/src/commandPalette/CommandList.tsx:419]: Side effect of running `yarn format`. Alongside the other `unrelated diffs` in there.

File: cmd/frontend/graphqlbackend/schema.graphql (hunk 19 above)
Comment (BolajiOlajide) [reply cmd/frontend/graphqlbackend/schema.graphql:1736]: I'll update.

Comment (eseliger) [reply cmd/frontend/graphqlbackend/schema.graphql:1736]: Please also capitalize Sourcegraph in the docstring :)
